### ISSUE TITLE
Redirect to `if let` and/or `match` when using unnecessary `.is_ok()` checks or similar

### ISSUE THREAD

**Issue by orowith2os** [rust-lang/rust-clippy#11377]
### What it does

When someone does an `.is_ok()` check on a `Result<T, E>` and proceeds to unwrap or otherwise take the value out of the enum, Clippy could suggest to use `if let`, especially if there's no `else` expression. 

Clippy may also be able to suggest `match` if there's an `else` expression.

Additionally, if someone uses `.map_err()` and unwraps everything else, a `match` statement might be a good suggestion to make.

### Advantage

- No duplicated checks to 1. see if a variable is okay and 2. unwrap it safely
- more standard Rust code
- more people get pushed to use other and nicer Rust features

### Drawbacks

A pain to implement, I figure. I can't really think of many cases where an `if x.is_ok() { x.unwrap() }` would be desired over an `if let`, or similar cases.

### Example

These won't be perfectly valid Rust code, but the idea is there, and is worth thinking about.

```rust
fn stuff(i: &str) {
    let x = File::open(i);
    if x.is_ok() {
        let y = x.unwrap();
        do_something_with(file);
    }
}
```

Could be written as:

```rust
fn stuff(i: &str) {
    let x = File::open(i);
    if let Ok(file) = x {
        do_something_with(file);
    }
}
```

```rust
fn stuff(i: &str) {
    let x = File::open(i);
    if x.is_ok() {
        let y = x.unwrap();
        do_something_with(file);
    } else {
        handle_error(x);
    }
}
```

Could be written as:

```rust
fn stuff(i: &str) {
    let x = File::open(i);
    match x {
        Ok(file) => do_something_with(file),
        Err(file) => handle_error(file),
    }
}
```

```rust
fn convert_paths_to_files(paths: Vec<String>) -> Vec<File> {
    paths
        .into_iter()
        .map(|path| {
            File::open(&path)
                .map_err(|err| {
                    eprintln!("Failed to open file '{}': {}", path, err);
                    std::process::exit(2);
                })
                .unwrap()
        })
        .collect()
}

```

Could be written as:

```rust
fn convert_paths_to_files(paths: Vec<String>) -> Vec<File> {
    paths
        .into_iter()
        .map(|path| {
            let x = File::open(&path);
            match x {
                Ok(file) => file,
                Err(e) => {
                    eprintln!("Failed to open file '{}': {}", path, err);
                    std::process::exit(2);
                }
            }
        })
        .collect()
}

```

**Comment by y21** [comment]
> When someone does an `.is_ok()` check on a `Result<T, E>` and proceeds to unwrap or otherwise take the value out of the enum, Clippy could suggest to use `if let`, especially if there's no `else` expression.

This one already exists: https://rust-lang.github.io/rust-clippy/master/index.html#/unnecessary_unwrap
That lint could probably be extended to catch `map_err` + `unwrap` too
